More arm fixes

mvwieringen opened the pull request:
This is a bit more controversial the trend on most platforms is to replace readdir_r with readdir given multi-threading is no longer a problem with readdir() . So I tried removing the usage of readdir_r from the 3 programs that still use it. It does cleanup things
a bit. fwflash still needs ses and scsi libraries which are not there so it cleaner but not working. Tested zonestat and that does
work and gives output for the global zone. I quoted in some commits the text from the manpage which seems to indicate this is the proper way of resolving things. I'm also fine with replacing the readdir_r from the two argument version to the 3 argument one and drop the two other patches.

richlowe commented on usr/src/cmd/rpcsvc/rpc.rusersd.c:
```diff
@@ -216,7 +216,7 @@ getutmpx_3(all, version, limit)
 			 * need to free this; done after svc_sendreply.
 			 */
-			*q = (struct utmpidle *)
+			*q = (struct utmpidle *)(uintptr_t)
```
You never have to cast the return of `malloc` `calloc` etc, they return `void *`.
Usually what this actually means is that someone didn't `#include <stddlib.h>`.  Casting through a longer type isn't going to help if the return is an implicit `int` as here.

mvwieringen replied:
Yeah already found this somewhat crazy that its needed didn't think about the obvious, fixed now

richlowe commented on usr/src/cmd/tip/hunt.c:
```diff
@@ -72,10 +72,10 @@ hunt(char *name)
 			(void) ioctl(FD, TCSETSF, &t);
 			(void) signal(SIGALRM, f);
-			return ((int)cp);
+			return ((intptr_t)cp);
```
The function here is still returning `int`.  I can't tell if you meant to fix that, or are relying on it implicitly truncating again.

mvwieringen replied:
Yeah sorry trying to work around some function that is looking crazy, fixed.

citrus-it replied:
From reading the call sites, I think this should actually be
```suggestion
			return (cp != NULL);
```

and the same at the end of the function, let's see if github will let me comment on that one

richlowe commented on usr/src/cmd/acct/acctprc.c:
```diff
@@ -156,7 +156,7 @@ void print_node(const void *node, VISIT order, int level) {
 
 	if (order == postorder || order == leaf) {
 		tb.ta_uid = (*(struct utab **)node)->ut_uid;
-		CPYN(tb.ta_name, (char *)uidtonam((*(struct utab **)node)->ut_uid));
```
This is probably another case where what's really going on is the an implicit declaration of `uidtonam` leading the compiler to think it returns `int`.  You don't need to cast through to the type it returns, but fix the implicit declaration.

You can probably catch a lot here by removing any `-Wno-implicit-function-declaration` from the makefiles.

mvwieringen replied:
Yeah its right in others but you are right missing a definition for uidtonam() added, fixed.

richlowe replied:
There's a _lot_ of implicit declaration warnings in here if you remove the entry from the Makefile to suppress them.
We should probably fix them all and switch to ANSI C to be extra sure everything is ok.  Without doing this, the compiler knows nothing about the types of parameters or returns and so can't tell us what's up

mvwieringen replied:
I will give it a try to see what it really needs this is really old and crufty code.

mvwieringen replied:
Hmm this is really a can of worms probably better to just drop this patch and one day either rwrite it from scratch or try fixing it if that is possible get the feeling this is far from 64 bits compatible.

richlowe commented on usr/src/cmd/mknod/mknod.c:
```diff
@@ -98,13 +98,13 @@ main(int argc, char **argv)
 		}
 		minno = (minor_t)number(argv[4]);
-		if (minno == (minor_t)-1 || minno > MAXMIN) {
+		if (minno == (minor_t)-1 || minno > (minor_t)MAXMIN) {
```
I know what's going on with this and MAXMAJ, but I'd like to talk to some people smarter than me about it, because this will allow mknod for the first time (ever?) to create devices that 32bit software can't interpret.

richlowe replied:
It would probably be best to change this to check against `MAXMAJ32` and `MAXMIN32` to ensure we don't create devices other software can't interpret, at least for now.  Rather than giving it the full range and possibly causing chaos for software trying to use the resulting nodes.

mvwieringen replied:
Ok done.

richlowe commented on usr/src/cmd/mailx/stralloc.c:
```diff
@@ -76,7 +76,7 @@ salloc(unsigned size)
 	s += 3;		/* needs alignment on quad boundary */
 	s &= ~03;
-#elif defined(i386)
+#elif defined(i386) || defined(__aarch64)
```
`__aarch64__` not `__aarch64`

mvwieringen replied:
Ok done.

richlowe commented:
For `fstyp` and `fwflash` it would be better to contribute those to illumos directly, since it's not related to ARM porting (they're using the 3-parameter version already.

richlowe commented on usr/src/cmd/zonestat/zonestatd/zonestatd.c:
```diff
@@ -4053,11 +4047,9 @@ zsd_open(zsd_ctl_t *ctl)
 	struct statvfs svfs;
 	int ret;
 	int i;
-	size_t size;
```
This is still used for eg at line 4102

mvwieringen replied:
Ah to savy on cleaning up, fixed.

mvwieringen commented:
> 

Ok dropped those will see if I can one day upstream them in some other way.

mvwieringen committed:
build 64-bit backup e.g. ufsdump/ufsrestore

mvwieringen committed:
build 64-bit zonestat

From the manpage:

While the readdir_r(function, was, originally, added, as, a, re-entrant,
version, of) readdir(), it is not recommended that readdir_r() be used in
new applications and existing software should instead use readdir(). The
readdir_r() function has been deprecated in many systems. Historically,
the data returned from readdir() was not specific to the directory stream
making it unsafe in a multi-threaded environment; however, that is no
longer the case.

mvwieringen committed:
build 64-bit cpio

mvwieringen committed:
fix 64 bit geniconvtbl

Added missing aarch64/Makefile

mvwieringen committed:
build 64-bit tar

mvwieringen committed:
build 64 bit mknod

mvwieringen committed:
build 64 bit acct

mvwieringen committed:
build 64 bit tip

mvwieringen committed:
build 64-bit mailx

citrus-it commented on usr/src/cmd/acct/acctcom.c:
```diff
@@ -556,7 +556,7 @@ println(void)
 #else
 	fprintf(stdout, " %-*.*s", OUTPUT_LSZ, OUTPUT_LSZ,
-	    a->ac_tty != (dev_t)-1? devtolin(a->ac_tty):"?");
+	    a->ac_tty != -1? devtolin(a->ac_tty):"?");
```
ac_tty is a `dev32_t` aka. `uint32_t` which is presumably why the original code cast -1 to `dev_t`. 
`dev_t` is a `ulong_t` which is of course larger in 64-bit, which is probably what the compiler complained about.

I suggest:
```suggestion
	    a->ac_tty != (dev32_t)-1 ? devtolin(a->ac_tty) : "?");
```

citrus-it commented on usr/src/cmd/acct/acctprc.c:
```diff
@@ -62,6 +62,7 @@ void **root = NULL;
 
 void output(void);
 void enter(struct ptmp *);
+char *uidtonam();
```
```suggestion
extern char *uidtonam(uid_t);
```
(I know the other files that you aren't touching use the K&R form, but let's fix this line here).

and now that the prototype is there, you can drop the explicit cast to `(char *)` where it is used, in `print_node()`.

citrus-it commented on usr/src/cmd/mailx/cmd1.c:
```diff
@@ -238,19 +238,19 @@ printhead(int mesg)
 	if ((mp->m_flag & (MREAD|MNEW)) == (MREAD|MNEW))
 		dispc = 'R';
-	if (!(int)value("bsdcompat") && (mp->m_flag & (MREAD|MNEW)) == MREAD)
+	if (!value("bsdcompat") && (mp->m_flag & (MREAD|MNEW)) == MREAD)
```
```suggestion
	if (value("bsdcompat") == NULL && (mp->m_flag & (MREAD|MNEW)) == MREAD)
```
Since `value()` returns a `char *` this should be an explicit comparison, and throughout.
The original (cast to int then boolean comparison) is awful!

citrus-it commented on usr/src/cmd/tar/tar.c:
```diff
@@ -8271,7 +8271,7 @@ read_xattr_hdr(attr_data_t **attrinfo)
 	if (link_len > 0)
 		xattr_linkp = (struct xattr_buf *)
-		    ((int)xattrp + (int)comp_len);
+		    ((uintptr_t)xattrp + (int)comp_len);
```
I think this should just be

```suggestion
    xattr_linkp = (struct xattr_buf *)(xattrp + comp_len);
```

citrus-it commented on usr/src/cmd/tip/hunt.c:
```diff
@@ -72,10 +72,10 @@ hunt(char *name)
 	}
 	(void) signal(SIGALRM, f);
-	return (deadfl ? -1 : (int)cp);
+	return (deadfl ? -1 : (intptr_t)cp);
```
```suggestion
	return (deadfl ? -1 : cp != NULL);
```

citrus-it commented on usr/src/cmd/tip/value.c:
```diff
@@ -152,7 +152,7 @@ vtoken(char *s)
 			cp++;
 			if (p->v_type&NUMBER)
-				vassign(p, (char *)atoi(cp));
+				vassign(p, (char *)(intptr_t)atoi(cp));
```
This is truly awful and ideally vassign() needs rewriting and the zzhack stuff ripping out, but this is no worse than what's already there and gets it building 64-bit, so..

citrus-it reviewed (changes requested):
Thanks for going through this @mvwieringen, there is some old and interesting code here.
I just have a few comments/changes - let me know if any of it doesn't make sense.

citrus-it commented on usr/src/cmd/mailx/collect.c:
```diff
@@ -121,7 +121,7 @@ collect(struct header *hp)
 	removefile(tempMail);
 
-	ignintr = (int)value("ignore");
+	ignintr = (intptr_t)value("ignore");
```
```suggestion
	ignintr = value("ignore") != NOSTR;
```

citrus-it commented on usr/src/cmd/mailx/collect.c:
```diff
@@ -438,7 +438,7 @@ collect(struct header *hp)
 				break;
 			}
-			grabh(hp, GMASK, (int)value("bsdcompat"));
+			grabh(hp, GMASK, (intptr_t)value("bsdcompat"));
```
```suggestion
			grabh(hp, GMASK, value("bsdcompat") != NOSTR);
```

citrus-it commented on usr/src/cmd/mailx/quit.c:
```diff
@@ -213,7 +213,7 @@ quit(
 	for (mp = &message[0]; mp < &message[msgCount]; mp++)
 		if (mp->m_flag & MBOX) {
-			if (msend(mp, obuf, (int)value("alwaysignore") ?
+			if (msend(mp, obuf, value("alwaysignore") == NOSTR ?
```
```suggestion
			if (msend(mp, obuf, value("alwaysignore") != NOSTR ?
```

citrus-it commented on usr/src/cmd/tip/hunt.c:
```diff
@@ -23,7 +23,7 @@ dead(void)
 }
 
-int
+intptr_t
```
You can drop this now and go back to `int`

citrus-it reviewed (changes requested):
Thanks, just a few stragglers.

richlowe commented:
Done.  Thanks for this!
